File: internal/githooks/githooks.go
package githooks

import (
	"fmt"
	"log"
	"sync"
)

type Linter struct {
	Tag    string
	Script func() error
}

func RunLinters(linters map[string]Linter) error {
	var mu sync.Mutex
	errs := map[string]error{}
	wg := new(sync.WaitGroup)

	for desc, linter := range linters {
		wg.Go(func() {
			log.Println(desc)
			if err := linter.Script(); err != nil {
				mu.Lock()
				errs[desc] = err
				mu.Unlock()
			}
		})
	}
	wg.Wait()

	if len(errs) > 0 {
		return fmt.Errorf("linter errors: %v", errs)
	}
	return nil
}

File: cmd-ocaml/githooks_lib/githooks.ml
type linter = {
  tag : string;
  script : unit -> (unit, string) result
}

let make_skip_checker () =
  let skip_env = try Sys.getenv "SKIP" with Not_found -> "" in
  let skips = String.split_on_char ',' skip_env |> List.map String.trim in
  fun tag -> List.mem tag skips

let run_linters linters should_skip =
  let running = ref [] in
  List.iter (fun (desc, linter) ->
    if not (should_skip linter.tag) then begin
      prerr_endline desc;
      let d = Domain.spawn (fun () ->
        match linter.script () with
        | Ok () -> None
        | Error e -> Some (desc, e)
      ) in
      running := d :: !running
    end
  ) linters;
  let errors =
    List.fold_left (fun acc d ->
      match Domain.join d with
      | None -> acc
      | Some err -> err :: acc
    ) [] !running
  in
  match errors with
  | [] -> Ok ()
  | errs -> Error errs
